service/history/hsm/tree_test.go

```diff
@@ -717,11 +796,214 @@ func TestNode_MultipleDeletedPaths(t *testing.T) {
 	err = branch2.DeleteChild(hsm.Key{Type: def1.Type(), ID: "b2child"})
 	require.NoError(t, err)
 
-	require.Empty(t, b1child.Outputs())
-	require.Empty(t, b2child.Outputs())
+	outputs, err := root.Outputs()
+	require.NoError(t, err)
+
+	// Verify both deletion operations exist
+	var foundB1ChildDel, foundB2ChildDel bool
+	for _, op := range outputs {
+		if del, ok := op.(hsm.DeleteOperation); ok {
+			if slices.Equal(del.Path(), b1child.Path()) {
+				foundB1ChildDel = true
+			}
+			if slices.Equal(del.Path(), b2child.Path()) {
+				foundB2ChildDel = true
+			}
+		}
+	}
+	require.True(t, foundB1ChildDel, "should find b1child deletion")
+	require.True(t, foundB2ChildDel, "should find b2child deletion")
+}
+
+func TestNode_PathPrefixEdgeCases(t *testing.T) {
+	root, err := hsm.NewRoot(reg, def1.Type(), hsmtest.NewData(hsmtest.State1), make(map[string]*persistencespb.StateMachineMap), &backend{})
+	require.NoError(t, err)
+
+	// Test similar paths at different levels:
+	// root
+	// ├── node1
+	// │   └── child1
+	// │       └── node1  (same name as parent)
+	// ├── node10
+	// └── node11
+
+	node1, err := root.AddChild(hsm.Key{Type: def1.Type(), ID: "node1"}, hsmtest.NewData(hsmtest.State1))
+	require.NoError(t, err)
+	node10, err := root.AddChild(hsm.Key{Type: def1.Type(), ID: "node10"}, hsmtest.NewData(hsmtest.State1))
+	require.NoError(t, err)
+	node11, err := root.AddChild(hsm.Key{Type: def1.Type(), ID: "node11"}, hsmtest.NewData(hsmtest.State1))
+	require.NoError(t, err)
+
+	// Add a path that reuses "node1" at a deeper level
+	child1, err := node1.AddChild(hsm.Key{Type: def1.Type(), ID: "child1"}, hsmtest.NewData(hsmtest.State1))
+	require.NoError(t, err)
+	deepNode1, err := child1.AddChild(hsm.Key{Type: def1.Type(), ID: "node1"}, hsmtest.NewData(hsmtest.State1))
+	require.NoError(t, err)
+
+	// Add transitions to all nodes
+	for _, node := range []*hsm.Node{node1, node10, node11, child1, deepNode1} {
+		err = hsm.MachineTransition(node, func(d *hsmtest.Data) (hsm.TransitionOutput, error) {
+			d.SetState(hsmtest.State2)
+			return hsm.TransitionOutput{}, nil
+		})
+		require.NoError(t, err)
+	}
+
+	err = root.DeleteChild(hsm.Key{Type: def1.Type(), ID: "node1"})
```

Comment:
nit: `node1.Key`.

Reply:
Are you saying you would prefer the `ID` to be `"node1.Key"`?